Add support for communication over SPI

jordan-powers opened the pull request.

jordan-powers committed:
Modify RobotLights.ino to use SPI instead of I2C

jordan-powers committed:
Update blinky-lights gradle version

jordan-powers committed:
Add SPI support to neopixel-leds-wpilibj

jordan-powers committed:
neopixel-leds v2.1

terbos90803 reviewed (commented):
Just a couple minor comments.  Thanks for getting on this.

lost1227 commented on Arduino/RobotLightsSPI/RobotLightsSPI.ino:
```diff
@@ -0,0 +1,244 @@
+// so we have to manually interface with the SPI hardware
+// See: https://arduino.stackexchange.com/questions/76528/spi-slave-programming-for-arduino
+ISR(SPI_STC_vect) {
+  volatile int nextQueueTail = (recvQueueTail + 1) % RECV_BUFF_SIZE;
```
Whoops, looks like I misunderstood what `volatile` means. 

I thought it would tell the compiler to keep the value in a register instead of memory, and that you had to do that with ISRs to keep the ISR as quick as possible.

Looks like it actually tells the compiler that the value might change at any point, and that the value should never be cached in a register.

I agree, `nextQueueTail` is not volatile. 

I should also probably update `recvBuff` to be volatile.

lost1227 commented on Arduino/RobotLightsSPI/RobotLightsSPI.ino:
```diff
@@ -0,0 +1,244 @@
+
+  strip.begin();
+
+  pinMode(MOSI, INPUT);
```
A fair bit. However, SPI uses dedicated hardware, so you can't adjust the pins in software unless you use software SPI, which has speed/throughput limitations.

On the arduino side, we need to make three changes:
1. Move the data wire from pin A4 (SDA) to pin 11 (MOSI)
2. Move the clock from pin A5 (SCL) to pin 13 (SCK)
3. Add a pulldown resistor (~10KΩ) between pin 10 (SS) and ground

Since we never send data from the LEDs to the rio, I don't think we need to bother connecting MISO. 

On the rio side, we basically need to cut off the i2c adapter, and solder MOSI to the data wire, SCL to the clock wire, and ground to ground. I've already bought a few connectors on amazon (<https://www.amazon.com/dp/B07SR3S2W3>), so  assuming I bought the right connectors, that should be pretty simple

lost1227 commented on Arduino/RobotLightsSPI/RobotLightsSPI.ino:
```diff
@@ -0,0 +1,244 @@
+
+  SPCR = 0;
+  SPCR |= _BV(SPE);
+  SPCR |= _BV(SPIE);
```
Yeah, that works.

jordan-powers committed:
Fix variable volatile lables

jordan-powers committed:
Coalesce SPCR bit assignments

jordan-powers committed:
Merge branch 'master' into spi

jordan-powers committed:
neopixel-leds v3.0

momentumrobotics-4999 committed:
Arduino SPI fixes

momentumrobotics-4999 committed:
SPI Fixes

jordan-powers committed:
Coalesce packet data in NeoPixels display before writing

jordan-powers committed:
Try to fix SPI endianness issues

It seems the rio is configured to send 16-bit words at a time.
However, the arduino is expecting 8-bit bytes at a time. This creates
an issue if the rio sends its words in little-endian, since the arduino
will see it as the bytes being sent in reverse order.
This patch (hopefully) fixes this issue.
